libpromises/locks.c

```diff
@@ -553,6 +517,60 @@ static void PromiseTypeString(char *dst, size_t dst_size, const Promise *pp)
     }
 }
 
+/**
+ * A helper best-effort function to prevent us from kill non CFEngine processes
+ * with matching PID-start_time combinations **when/where its easy to check**.
+ */
+static bool IsCfengineLockHolder(pid_t pid)
+{
+    char procfile[PATH_MAX];
+    snprintf(procfile, PATH_MAX, "/proc/%ju/comm", (uintmax_t) pid);
+    int f = open(procfile, O_RDONLY);
+    /* On platforms where /proc doesn't exist, we would have to do a more
+       complicated check probably not worth it in this helper best-effort
+       function. */
+    if (f == -1)
+    {
+        /* assume true where we cannot check */
+        return true;
+    }
+
+    /* more than any possible CFEngine lock holder's name */
+    char command[32] = {0};
+    ssize_t n_read = FullRead(f, command, sizeof(command));
+    close(f);
+    if ((n_read <= 1) || (n_read == sizeof(command)))
+    {
+        Log(LOG_LEVEL_VERBOSE, "Failed to get command for process %ju", (uintmax_t) pid);
+        /* assume true where we cannot check */
+        return true;
+    }
+    if (command[n_read - 1] == '\n')
+    {
+        command[n_read - 1] = '\0';
+    }
+
+    /* potential CFEngine lock holders */
+    const char * const cfengine_procs[] = {
+        "lt-cf-agent",  /* when running from a build with 'libtool --mode=execute' */
+        "cf-agent",
+        "cf-execd",
+        "cf-serverd",
+        "cf-monitord",
+        "cf-hub",
+        "cf-reactor",
+        NULL
+    };
+    for (const char * const *cfe_proc = cfengine_procs; *cfe_proc != NULL; cfe_proc++)
+    {
+        if (StringEqual(*cfe_proc, command))
```

Comment:
Doesn't the command usually have a bunch of arguments, which would make this not equal?

Reply:
There's a difference between `/proc/$PID/cmdline` and `/proc/$PID/comm`. The latter doesn't contain the arguments.